app/Filament/Resources/UserResource/Pages/Roles.php

```diff
@@ -37,16 +43,60 @@ public function table(Table $table): Table
                 Tables\Actions\AttachAction::make()
                     ->label(__('Add'))
                     ->color('primary')
-                    ->authorize(fn () => auth()->user()->can('updateRoles', $this->getRecord()))
+                    ->authorize(fn () => $user->can('updateRoles', $this->getRecord()))
                     ->recordTitle(fn (Model $record) => __('permission.role.' . $record->name))
                     ->preloadRecordSelect()
-                    ->recordSelectOptionsQuery(fn (Builder $query) => $query->whereIn('name', auth()->user()->assignableRoles)),
+                    ->recordSelectOptionsQuery(fn (Builder $query) => $query->whereIn('name', $user->assignableRoles))
+                    ->after(function ($data) {
+                        /** @var User $targetUser */
+                        $targetUser = $this->getRecord();
+
+                        $attachedRole = Role::findById((int) $data['recordId']);
+
+                        if ($attachedRole->name === Role::DEVELOPER_JUNIOR) {
+                            $targetUser->removeRole(Role::DEVELOPER);
+                            $targetUser->removeRole(Role::DEVELOPER_STAFF);
+                            $targetUser->removeRole(Role::DEVELOPER_VETERAN);
+
+                            $targetUser->setAttribute('Permissions', Permissions::JuniorDeveloper);
+                            $targetUser->save();
+                        } elseif ($attachedRole->name === Role::DEVELOPER) {
+                            $targetUser->removeRole(Role::DEVELOPER_JUNIOR);
+                            $targetUser->removeRole(Role::DEVELOPER_STAFF);
+                            $targetUser->removeRole(Role::DEVELOPER_VETERAN);
+
+                            $targetUser->setAttribute('Permissions', Permissions::Developer);
+                            $targetUser->save();
+                        } elseif ($attachedRole->name === Role::DEVELOPER_STAFF) {
+                            $targetUser->removeRole(Role::DEVELOPER_JUNIOR);
+                            $targetUser->removeRole(Role::DEVELOPER);
+                            $targetUser->removeRole(Role::DEVELOPER_VETERAN);
+
+                            $targetUser->setAttribute('Permissions', Permissions::Developer);
+                            $targetUser->save();
+                        } elseif ($attachedRole->name === Role::DEVELOPER_VETERAN) {
+                            $targetUser->removeRole(Role::DEVELOPER_JUNIOR);
+                            $targetUser->removeRole(Role::DEVELOPER);
+                            $targetUser->removeRole(Role::DEVELOPER_STAFF);
+
+                            $targetUser->setAttribute('Permissions', Permissions::Registered);
```

Comment:
What is "Developer Veteran"? Based on the implementation, I recommend renaming it to "Retired Developer" or something similar.